docs/pipe-protocol.md

```diff
@@ -0,0 +1,270 @@
+- Open question: For "unknown" messages? What is the best to do to preserve the max possible compatibility?
+  - Today: when .NET CLI receives an "unknown serializer id", it skips reading this message, and it responds with VoidResponse.
+  - If the expected response of this message on MTP side is not VoidResponse, then MTP will fail with InvalidCastException.
+  - We might have two different scenarios here:
+    - A future MTP request **really** requires a response, and MTP cannot move forward without a response. In this case, we cannot do anything. It's a breaking change.
+    - A future MTP request might not really require a response. In this case, VoidResponse can be a meaning of ".NET CLI didn't recognize the request".
+    - However, this brings ambiguity. If a future MTP message isn't recognized by .NET CLI, and MTP is already expecting VoidResponse.
+      - There is no way to tell if .NET CLI understood our message or not.
+      - Could it be important for MTP to know whether or not .NET CLI understood the request?
+      - In this case, should we have a "special" response that means "I didn't understand your request"?
+      - I think such special response gives us the most flexibility.
+
+## General protocol message format
+
+- 4 bytes: message size
+- 4 bytes: serializer id (indicator of the message type)
+- x bytes: message payload
+  - property/field count (2 bytes)
+  - for each property/field:
+    - property/field id (2 bytes)
+    - property/field size (4 bytes)
+    - property/field value (n bytes)
+      - If the value is an "array":
+        - array length (4 bytes)
+        - each array element follows again the message payload format (id, size, and value).
+
+```mermaid
+graph TD
+    A[Message] --> B["Message Size (4 bytes)"]
+    A --> C["Serializer ID (4 bytes)"]
+    A --> D["Message Payload (x bytes)"]
+
+    D --> D1["Property/Field Count (2 bytes)"]
+    D --> D2[Property/Field]
+
+    D2 --> D2a["Property/Field ID (2 bytes)"]
+    D2 --> D2b["Property/Field Size (4 bytes)"]
+    D2 --> D2c["Property/Field Value (n bytes)"]
+
+    D2c --> E{Is Array?}
+    E -->|Yes| F["Array Length (4 bytes)"]
+    F --> G[Array Element]
+    G --> G1["Element ID (2 bytes)"]
+    G --> G2["Element Size (4 bytes)"]
+    G --> G3["Element Value (n bytes)"]
+```
+
+## Future considerations
+
+- Youssef: Can we get rid of InstanceId completely? I personally don't believe it was the right design.
+  - If we want to simply track retries, that could simply have been a metadata on test results.
+  - Having it as a metadata on test results even allows for in-process retries to be reported.
+  - Unfortunately, removing InstanceId is going to be a breaking change in the protocol.
+  - What we could do?
+    - Keep InstanceId for compatibility.
+    - Add the metadata on test results as "another" way of knowing retries.
+    - In handshake request, tell .NET CLI that we have the "new" way of knowing retries.
```

Comment:
are we considering using the VS server protocol for `dotnet test` at any point? we already have capabilities there and this would make it easier to include features such as retries to both CLI and VS/VsCode simulatenously

Reply:
I personally would consider the opposite :smile:
Any kind of "capability" can potentially be added as a property in handshake request.